lightly/api/api_workflow_datasources.py

```diff
@@ -51,44 +66,60 @@ def _download_raw_files(
                 cursor=cursor,
                 use_redirected_read_url=use_redirected_read_url,
                 **relevant_filenames_kwargs,
-                **kwargs
+                **kwargs,
             )
             cursor = response.cursor
             samples.extend(response.data)
             if progress_bar is not None:
                 progress_bar.update(len(response.data))
-        samples = [(s.file_name, s.read_url) for s in samples]
+        sample_map = {}
+        for s in samples:
+            if re.match("^\.*/", s.file_name):
+                raise ValueError(
+                    f"Absolute file paths like {s.file_name} are not supported"
+                    " in relevant filenames file {relevant_filenames_file_name}"
+                )
```

Comment:
I am unsure why this should not be allowed? And if it's not allowed, why wouldn't we do the check in the API?

Reply:
R.e. location of check, it could be done in either place and having it in the Lightly package seems reasonable if we might ever call this directly/ not via the API (the error can be propagated as appropriate).

R.e. `../` we don't know the parent directory so can't support it.

R.e. `./filename` we could treat that as equivalent to `filename` if desired/ expected.

R.e. `/filename`, this is invalid since filenames are meant to be relative to the bucket. We could strip the leading `/` and try to proceed, or we can error out as in this PR and request valid input.

Reply:
R.e. `../` I fully agree
R.e. `./filename`. I would treat it just like `../`. It is just an extra way of doing the same thing, which means more maintenance for us. I would rather error than implement support for it.
R.e. `/filename`. This is a plain error, also error. Stripping and trying will just lead to failures in the corruptness check the user does not understand. Better fail early with a good error message.

Reply:
Regarding checking it in the API: 
- Yes, that is possible
- it does not allow for so nice error messages.
- it does allow to reuse some logic if we use e.g. relevant filenames with local files

Reply:
The logic for local files would be different. Having a `../` or a `./` or even a `/home/blabla` would be totally legit when dealing with local files as its the users own filesystem. So the logic can't be shared.

I don't think the python client should have anything to say in this regard. Especially since `.` and `/` do not represent folder structures in a blob store (bucket) but just some arbitrary character in the name of a blob. We only need the name here for ensuring that the file only exists once. And thats what this issue is about.

Reply:
> The logic for local files would be different. Having a ../ or a ./ or even a /home/blabla would be totally legit when dealing with local files as its the users own filesystem.

For local files, we would still disallow it, as we don't want to support extra logic to handle it. E.g. we don't want to check it everywhere if the file is absolute or not and thus if we need to append the path or not.

Reply:
> I don't think the python client should have anything to say in this regard. Especially since . and / do not represent folder structures in a blob store (bucket) but just some arbitrary character in the name of a blob.

I am not sure what we are discussing here:
- Are you against disallowing filenames with ./ and ../?
- Or are your for disallowing them, but want the same logic implemented in the API, not the pip package?

Reply:
I don't have any strong view here, and I'd really like to get this done!

The `_DatasourcesMixin` method `_download_raw_files` is only intended as applicable to remote datasources. It cannot be used for local datasources. Therefore all filenames are paths relative to the bucket/ root of the remote datasource. A filename beginning with `/` or `../` is clearly *always* invalid. The `./` case is debatable (though it would be redundant, there's no use case to support it, and we would rather treat it as invalid to avoid having to make changes elsewhere in the code).

In particular: we have had cases of a client passing in absolute file paths, which was handled silently and then caused failure during corruptness check which was a terrible user experience!

We want to prevent `_DatasourcesMixin._download_raw_files` ever being called with invalid input.

We could have the API catch this and return an error. That would work wherever `_DatasourcesMixin._download_raw_files` is called via the API.

However, at least in theory, it is possible to call `_DatasourcesMixin._download_raw_files` locally (not via the API), with a remote datasource, with invalid filenames. Even in that case it would be better to throw a useful error & informative error message. The present implementation does that, and also keeps API users happy.

So, isn't it best to stick with this? Or is there a good reason this would be better handled at the API level?

Reply:
I would merge it like this here. We can move it to the API anytime later if we see the need for it.

Reply:
So to clarify:
I do **not** want to have this logic. Not in the API and not in the client.

> A filename beginning with / or ../ is clearly always invalid

This is not true. Even though a `/` in the key name is stupid it still is valid. `/` is just used as a delimiter in the UI. See here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

If someone has a basepath of `my-bucket/my-stuff/` and they put `../bla.jpg` in to the relevantFilenames file, then it will try to retrieve `my-bucket/my-stuff/../bla.jpg`. Yes, its stupid but we should not assume anything about the clients naming strategy.

> In particular: we have had cases of a client passing in absolute file paths, which was handled silently and then caused failure during corruptness check which was a terrible user experience!

I agree that this is bad UX. But to error out seems like bad UX as well. I'd rather add some "verify"/"dry run" where we don't try to load ALL the data (e.g only list until we have 100 images) and move on. And to also add examples to docs of wrong examples (which should be added in this PR at least).

This whole debate came from something that is not related to the issue at hand. If you both think this is the right thing to do then I will step back but I want to make it clear that I don't agree.